### PR TITLE
Implement sequence syntax

### LINKED ISSUE #168: Add Syntax for sequences of types and expressions
From https://github.com/rise-lang/shine/pull/167#pullrequestreview-650983593:
> The {} parameters should define a family of primitives and therefore not contain language constructs like is done for ForeignFunctionCall, MakeArray, HostExecution, etc to deal with sequences and maps of types/expressions. Instead we should come up with a syntax to move such sequences to the regular () parameters.

### PR DESCRIPTION
**Pull request by michel-steuwer**
This implements a syntax for sequences of arguments for primitives and resolves #168

### CODE REVIEW

File: src/main/scala/shine/DPIA/Data.scala [hunk 1]
```diff
@@ -14,6 +14,8 @@ final case class IntData(i: Int) extends Data(int) {
   override def toString: String = i.toString
 }
 
+final case class NatAsIntData(n: Nat) extends Data(int)
```
Comment (Bastacyclop) [inline src/main/scala/shine/DPIA/Data.scala:17]: Is there a reason to introduce this change in this PR, or is it orthogonal? Does this not require changes to `VisitAndRebuild` and others? Maybe we need to think about cleaning up `Literal(NatData(_))`, `Natural(_)` and other constructs at some point.

File: src/main/scala/shine/OpenCL/Compilation/SeparateHostAndKernelCode.scala [hunk 2]
```diff
@@ -59,7 +60,7 @@ object SeparateHostAndKernelCode {
       freeNats match {
         case v +: rest => iterNats(
           DepLambda[NatKind](NatIdentifier(v.name, v.range))(definition),
-          Natural(v) +: args, rest)
+          Literal(NatAsIntData(v)) +: args, rest)
```
Comment (Bastacyclop) [inline src/main/scala/shine/OpenCL/Compilation/SeparateHostAndKernelCode.scala:63]: so now the argument is an `int` instead of a `nat`, which is in the end going to be the same at runtime, but why this change?

**Review by Bastacyclop — COMMENTED** [review]
The end result looks good to me :+1: However the code in the meta package is getting quite cryptic for me, I wonder if some high-level comments would help future maintenance.

File: meta/src/main/scala/meta/generator/DPIAPrimitives.scala [hunk 3]
```diff
@@ -122,25 +127,102 @@ ${generateCaseClass(Type.Name(name), scalaParamsString, params, returnType)}
...
+          ids2 <- getUnrolledIds(outT)
+        } yield ids1 concat ids2
+      case AST.DepFunType(_, _, t) =>
+        getUnrolledIds(t)
+      case AST.Identifier(_) => None
+      case AST.VariadicType(n, ty) => throw new Exception("This function should not be called on a variadic type")
+    }
+  }
+
+  def getUnrolledIds(typeAST: rise.Type.AST): Option[Seq[rise.Type.AST.UnrolledIdentifier]] = {
```
[the remark this answers is not in the document]
Comment (michel-steuwer) [reply meta/src/main/scala/meta/generator/DPIAPrimitives.scala:190]: Will do

File: src/main/scala/shine/DPIA/fromRise.scala [hunk 4]
```diff
@@ -673,31 +673,32 @@ object fromRise {
...
-            fun[ExpType](ExpType(inTs(i), read), a =>
-              buildFFCall(args :+ a))
-          } else {
-            ForeignFunctionCall(decl, inTs, args)(outT)
-          }
+        assert(inTs.length == n)
+
+        inTs.foldRight[Phrase[_ <: PhraseType]](
+          depFun[DataKind](outT)({
+            val args = Seq.tabulate(n)(i => Identifier(freshName("x"), ExpType(inTs(i), read)))
```
[the remark this answers is not in the document]
Comment (michel-steuwer) [reply src/main/scala/shine/DPIA/fromRise.scala:692]: The `fun` syntax creates a scala binding that has to be scoped appropriately and I think this would have made the implementation much more complicated.

File: src/main/scala/shine/DPIA/fromRise.scala [hunk 5]
```diff
@@ -673,31 +673,32 @@ object fromRise {
...
-        def buildFFCall(args: Vector[Phrase[ExpType]]
-                            ): Phrase[_ <: PhraseType] = {
-          val i = args.length
-          if (i < inTs.length) {
-            fun[ExpType](ExpType(inTs(i), read), a =>
-              buildFFCall(args :+ a))
-          } else {
-            ForeignFunctionCall(decl, inTs, args)(outT)
-          }
+        assert(inTs.length == n)
```
[the remark this answers is not in the document]
Comment (michel-steuwer) [reply src/main/scala/shine/DPIA/fromRise.scala:688]: I will check if this is checked as part of the well-formedness check (or if it can be done there).

File: src/main/scala/shine/DPIA/Data.scala (hunk 1 above)
Comment (michel-steuwer) [reply src/main/scala/shine/DPIA/Data.scala:17]: This and your next comment are linked:

adding the explicit type information to `KernelCallCmd` has exposed this "bug" that the type we record (`exp[int, read]`) and the argument value we create (a `Natural`) did not match.
This was not a problem before, as (as you say in the commend below) "is going to be the same at runtime".

To bring the types and arguments into line, I added this way to add a literal with a natural number whose type is an int. This is obviously not ideal.

I will see if I can make the opposite change (i.e. change the type rather than the value), but I believe that had greater implications elsewhere ...

I agree that we should at some point revisit our `Literal`s, `Natural`, etc.

Comment (michel-steuwer) [reply src/main/scala/shine/DPIA/Data.scala:17]: I created an issue to track this: #172

File: src/main/scala/shine/DPIA/primitives/functional/ForeignFunctionCall.scala [hunk 6]
```diff
@@ -7,8 +7,14 @@ import shine.DPIA.Phrases._
...
 import shine.DPIA._
-final case class ForeignFunctionCall(funDecl: rise.core.ForeignFunction.Decl, inTs: Seq[DataType], args: Seq[Phrase[ExpType]])(val outT: DataType) extends ExpPrimitive {
-  {}
+final case class ForeignFunctionCall(funDecl: rise.core.ForeignFunction.Decl, n: Int)(val inTs: Seq[DataType], val outT: DataType, val args: Seq[Phrase[ExpType]]) extends ExpPrimitive {
+  {
+    args.zip(inTs).foreach({
+      case (args, inTs) =>
+        args :: expT(inTs, read)
+    })
+  }
```
[the remark this answers is not in the document]
Comment (Bastacyclop) [reply src/main/scala/shine/DPIA/primitives/functional/ForeignFunctionCall.scala:16]: As well as `assert(args.length == n)`.

[the remark this answers is not in the document]
Comment (Bastacyclop) [reply src/main/scala/shine/DPIA/primitives/functional/ForeignFunctionCall.scala:16]: assert {
  // checks
  true
}

**Review by Bastacyclop — APPROVED** [review]
:+1:
